psl-core/src/main/java/org/linqs/psl/util/VizDataCollection.java

```diff
@@ -0,0 +1,233 @@
+package org.linqs.psl.util;
+
+import org.linqs.psl.grounding.GroundRuleStore;
+import org.linqs.psl.model.atom.Atom;
+import org.linqs.psl.model.atom.GroundAtom;
+import org.linqs.psl.model.rule.GroundRule;
+import org.linqs.psl.model.rule.Rule;
+import org.linqs.psl.model.rule.UnweightedGroundRule;
+import org.linqs.psl.model.rule.WeightedGroundRule;
+import org.linqs.psl.model.rule.logical.AbstractLogicalRule;
+import org.linqs.psl.model.term.Constant;
+import org.linqs.psl.model.term.Variable;
+
+import org.json.JSONArray;
+import org.json.JSONObject;
+
+import java.util.ArrayList;
+import java.io.File;
+import java.io.FileWriter;
+import java.util.HashMap;
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.HashSet;
+
+public class VizDataCollection {
+
+    private static VizDataCollection vizData = null;
+    private static Runtime runtime = null;
+
+    private JSONObject fullJSON;
+    private JSONArray predictionTruthArray;
+    private JSONArray ruleCountArray;
+    private JSONArray totRuleSatArray;
+    private JSONArray violatedGroundRulesArray;
+
+    JSONObject rules;
+    JSONObject groundRules;
+    JSONObject groundAtoms;
+
+    public static ArrayList<GroundRule> violatedGroundRulesList = new ArrayList<>();
+
+    static {
+        init();
+    }
+
+    // Static only.
+    private VizDataCollection() {
+        fullJSON = new JSONObject();
+        predictionTruthArray = new JSONArray();
+        ruleCountArray = new JSONArray();
+        totRuleSatArray = new JSONArray();
+        violatedGroundRulesArray = new JSONArray();
+        rules = new JSONObject();
+        groundRules = new JSONObject();
+        groundAtoms = new JSONObject();
+    }
+
+    private static synchronized void init() {
+        if (runtime != null) {
+            return;
+        }
+        vizData = new VizDataCollection();
+        runtime = Runtime.getRuntime();
+        runtime.addShutdownHook(new ShutdownHook());
+    }
+
+    //We want to make:
+    // A jsonObject filled with JSONArrays
+    // Will be organized into different JSON arrays that refer to specific modules, all in one object
+    //e.x.
+        // [
+        //     {predicate: Friends((bob,george), prediction: 0.00003, truth: 1}
+        //     {predicate: Friends((alice,george), prediction: 0.00003, truth: 1}
+        //     etc...
+        // ]
+    public static void outputJSON() {
```

Comment:
You can output a full object to JSON, instead of each thing individually:
http://stleary.github.io/JSON-java/org/json/JSONObject.html#JSONObject-java.lang.Object-java.lang.String:A-

The plan here would be to create a private static inner class with all public members.
(It is fine to have public members since the entire class is private.)
Here is where you can collect all the data (replacing most of your other containers).
Then here in output, we should just be able to output that one object.
It should also make it easier to organize the keys in the JSON.

Also, you don't want to carry around JSON objects if you can avoid it, they are much less usable than normal Java types.
You only want to have JSON objects right when you are generating output.
So ideally your datatypes will be things like List and Map instead of JSONArray and JSONObject (until it is time to output).

Reply:
Resolved